src/Analyzer/Passes/FuseFunctionsPass.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Analyzer/Passes/FuseFunctionsPass.h>
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Functions/FunctionFactory.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class FuseFunctionsMatcher : public InDepthQueryTreeVisitor<FuseFunctionsMatcher>
+{
+public:
+
+    static bool matchFunctionName(const String & name)
+    {
+        return name == "count" || name == "sum" || name == "avg";
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !function_node->isAggregateFunction() || !matchFunctionName(function_node->getFunctionName()))
+            return;
+
+        auto argument_hash = function_node->getArgumentsNode()->getTreeHash();
+        mapping[argument_hash].push_back(&node);
+    }
+
+    struct QueryTreeHashForMap
+    {
+        size_t operator()(const IQueryTreeNode::Hash & hash) const { return hash.first ^ hash.second; }
+    };
+
+    /// argument -> list of sum/count/avg functions with this argument
+    std::unordered_map<IQueryTreeNode::Hash, std::vector<QueryTreeNodePtr *>, QueryTreeHashForMap> mapping;
```

Comment:
This is not valid mapping, because it will not use query tree node `equalTo` method for comparison. I would prefer not to rely that if `IQueryTreeNode::Hash` is the same, it is the same node.
To make this work, we can create structure with IQueryTreeNode hash and IQueryTreeNode pointer, provide special comparator in map, and it will work.

Reply:
I updated `OrderByLimitByDuplicateEliminationPass` implementation with my comment. https://github.com/ClickHouse/ClickHouse/pull/42704/files#diff-339c03f58cda732b6de178b7d1c02d81283c20a40039b09160eef37831d1d45cR13